**imports**

In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
import seaborn as sns
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split

### 1.load data and preprocessing

1.1 load_data

In [16]:
#loading the data
(x_train_full, y_train_full), (x_test, y_test) = mnist.load_data()

In [17]:
from sklearn.model_selection import train_test_split

x_train, x_valid, y_train, y_valid = train_test_split(x_train_full, y_train_full, random_state=42)

1.2 preprocessing

In [21]:
# Reshape training data
x_train = x_train.reshape(45000, 784).astype('float32')

# Reshape validation data
x_test = x_test.reshape(10000, 784).astype('float32')

# Normalize training data
x_train /= 255

# Normalize validation data
x_test /= 255

# Define number of classes
n_classes = 10

In [22]:
y_train

array([0, 1, 1, ..., 0, 0, 4], dtype=uint8)

### 2.Baseline Model with Traditional Machine Learning Algorithms

2.1 implementing a logistic regression algorithm

In [23]:
from sklearn.linear_model import LogisticRegression

In [30]:
log_reg = LogisticRegression(random_state=42)

In [31]:
log_reg.fit(x_train, y_train)

C:\Users\Mo\.conda\envs\ml\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

2.2 Evaluate the model performance

In [32]:
log_reg.score(x_test, y_test)

0.9241

Logistic Regression Score: **92.4%**

### 3. Ensemble of Machine Learning Algorithms

In [37]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.svm import SVC

models = [
    ('a', DecisionTreeClassifier(random_state=42)),
    ('b', RandomForestClassifier(random_state=42)),
    ('c', SVC(random_state=42))
]

voting_clf = VotingClassifier(models, n_jobs=-1)

In [38]:
voting_clf.fit(x_train, y_train)

TypeError: VotingClassifier.fit() got an unexpected keyword argument 'n_jobs'

In [36]:
voting_clf.score(x_test, y_test)

0.9699

Voting Ensemble Score: 97%

### 4. Neural Network Model Design

In [104]:

(x_train_full, y_train_full), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train, x_valid, y_train, y_valid = train_test_split(x_train_full, y_train_full, random_state=42)

In [105]:
x_train = x_train.astype('float64')
x_test = x_test.astype('float64')
x_valid = x_valid.astype('float64')

In [106]:
x_train /= 255.0
x_test /= 255.0
x_valid /= 255.0

In [107]:
x_train.shape

(45000, 28, 28)

In [122]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(300, activation='relu'),
    tf.keras.layers.Dense(100, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)
model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.fit(x_train, y_train, epochs=10, validation_data=(x_valid, y_valid))

Epoch 1/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8806 - loss: 0.4048 - val_accuracy: 0.9568 - val_loss: 0.1443
Epoch 2/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9703 - loss: 0.0975 - val_accuracy: 0.9717 - val_loss: 0.0942
Epoch 3/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9809 - loss: 0.0610 - val_accuracy: 0.9699 - val_loss: 0.0943
Epoch 4/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9862 - loss: 0.0409 - val_accuracy: 0.9735 - val_loss: 0.0920
Epoch 5/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9893 - loss: 0.0334 - val_accuracy: 0.9766 - val_loss: 0.0840
Epoch 6/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9916 - loss: 0.0269 - val_accuracy: 0.9727 - val_loss: 0.1071
Epoch 7/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9929 - loss: 0.0210 - val_accuracy: 0.9764 - val_loss: 0.0991
Epoch 8/10
1407/1407 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9937 - loss: 0.0182 - 

In [123]:
model.evaluate(x_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 599us/step - accuracy: 0.9702 - loss: 0.1278


[0.11115114390850067, 0.9746000170707703]

In [124]:
model.evaluate(x_train, y_train)

1407/1407 ━━━━━━━━━━━━━━━━━━━━ 1s 621us/step - accuracy: 0.9950 - loss: 0.0159


[0.015660259872674942, 0.9949555397033691]

2 hidden layers, 300, 100 neurons NN Evaluation test: 97.8% train: 99.6% \n
3 hidden layers, 300, 300, 100 neurons NN Evaluation test: 94.8% train: 97.9%


In [126]:
import keras_tuner as kt


def build_model(hp, batch=False):
    with tf.device('/device:gpu:0'):
        n_hidden = hp.Int("n_hidden", min_value=0, max_value=5, default=2)
        n_neurons = hp.Int("n_neurons", min_value=16, max_value=280)
        learning_rate = hp.Float("learning_rate", min_value=1e-4, max_value=1e-2,
                                 sampling="log")

        optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate, beta_1=0.9,
                                             beta_2=0.999, epsilon=1e-8)

        model = tf.keras.Sequential()

        model.add(tf.keras.layers.Flatten(input_shape=(28, 28)))
        if batch:
            model.add(tf.keras.layers.BatchNormalization())
        for _ in range(n_hidden):
            model.add(tf.keras.layers.Dense(n_neurons, activation="relu"))

        model.add(tf.keras.layers.Dense(10, activation='softmax'))
        model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=["accuracy"])
        return model

In [127]:
from keras_tuner import GridSearch, RandomSearch

random_search_tuner = kt.RandomSearch(
    build_model, objective="val_accuracy", max_trials=3, overwrite=True,
    directory="trails_list", project_name="my_rnd_search", seed=42)

random_search_tuner.search(x_train, y_train, epochs=10,
                           validation_data=(x_valid, y_valid))


Trial 5 Complete [00h 00m 26s]
val_accuracy: 0.9564666748046875

Best val_accuracy So Far: 0.9802666902542114
Total elapsed time: 00h 02m 41s


In [128]:
best_trial = random_search_tuner.oracle.get_best_trials(num_trials=1)[0]
best_trial.summary()

Trial 1 summary
Hyperparameters:
n_hidden: 1
n_neurons: 251
learning_rate: 0.000505466655750672
Score: 0.9802666902542114
